src/types.ts

```diff
@@ -0,0 +1,174 @@
+/**
+ * @fileoverview Internal types used inside htmx.js but are not part of
+ * the public API. The main benefit of having this file is that you can
+ * have better DX defining types in TypeScript and then just import them
+ * into the JsDoc comments in the htmx.js file. */
+
+interface HtmxTriggerSpecification {
+  sseEvent?: string;
+  trigger: string;
+  root?: Element | Document | null;
+  threshold?: string;
+  delay?: number;
+  pollInterval?: number;
+}
+
+export interface HtmxSwapSpecification {
+  swapStyle: SwapStyle;
+  swapDelay: number;
+  settleDelay: number;
+
+  scroll?: "top" | "bottom";
+  scrollTarget?: string | null;
+
+  show?: "top" | "bottom";
+  showTarget?: string | null;
+
+  transition?: boolean;
+  ignoreTitle?: boolean;
+  focusScroll?: boolean;
+}
+
+interface HtmxSettleInfo {
+  title?: string;
+  tasks?: (() => void)[];
+  elts?: Element[];
+}
+
+interface ListenerInfo {
+  listener: EventListener;
+  on: HTMLElement;
+  trigger: string;
+}
+
+/** the http verb used in the request (lowercase) */
+type Verb = "get" | "post" | "put" | "delete" | "patch";
+
+interface KnownInternalData {
+  initHash?: number | null;
+  listenerInfos?: ListenerInfo[];
+  path?: string;
+  streamPaused?: boolean;
+  streamReader?: ReadableStreamDefaultReader;
+  verb?: Verb;
+  lastButtonClicked?: Element | null;
+  timeout?: number;
+  webSocket?: WebSocket;
+  sseEventSource?: EventSource;
+  onHandlers?: { event: string; listener: EventListener }[];
+  xhr?: XMLHttpRequest;
+  requestCount?: number;
+}
+
+type InternalData = KnownInternalData & Record<PropertyKey, any>;
+
+interface InputValues {
+  errors: any[];
+  values: Record<string, string>;
+}
+
+interface Pollable {
+  polling: boolean;
+}
+
+interface TriggerHandler {
+  (elt: Element, evt: Event): void;
+  (): void;
+}
+
+export type SwapStyle =
+  | "innerHTML"
+  | "outerHTML"
+  | "beforebegin"
+  | "afterbegin"
+  | "beforeend"
+  | "afterend"
+  | "delete"
+  | "none";
+
```

Comment:
we should probably add `(string & NonNullable<unknown>)` like this:
```ts
export type SwapStyle =
    | "innerHTML"
    | "outerHTML"
    | "afterbegin"
    | "beforebegin"
    | "afterend"
    | "beforeend"
    | "none"
    | "delete"
    | (string & NonNullable<unknown>);
```

because there are other swap styles (like `morph:innerHTML` when using idiomorph)

this way, editor autocomplete will still be there for known values, but any string *can* be passed.